debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,599 @@
+import org.bson.Document;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.MongoDatabase;
+
+import io.debezium.config.ConfigurationDefaults;
+import io.debezium.connector.mongodb.ConnectionContext.MongoPrimary;
+import io.debezium.pipeline.ErrorHandler;
+import io.debezium.pipeline.EventDispatcher;
+import io.debezium.pipeline.EventDispatcher.SnapshotReceiver;
+import io.debezium.pipeline.source.spi.SnapshotChangeEventSource;
+import io.debezium.pipeline.source.spi.SnapshotProgressListener;
+import io.debezium.pipeline.spi.ChangeRecordEmitter;
+import io.debezium.pipeline.spi.OffsetContext;
+import io.debezium.pipeline.spi.SnapshotResult;
+import io.debezium.pipeline.txmetadata.TransactionContext;
+import io.debezium.util.Clock;
+import io.debezium.util.Metronome;
+import io.debezium.util.Strings;
+import io.debezium.util.Threads;
+import io.debezium.util.Threads.Timer;
+
+/**
+ * A {@link SnapshotChangeEventSource} that performs multi-threaded snapshots of replica sets.
+ *
+ * @author Chris Cranford
+ */
+public class MongoDbSnapshotChangeEventSource implements SnapshotChangeEventSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbSnapshotChangeEventSource.class);
+
+    private static final String AUTHORIZATION_FAILURE_MESSAGE = "Command failed with error 13";
+
+    private final MongoDbConnectorConfig connectorConfig;
+    private final MongoDbTaskContext taskContext;
+    private final MongoDbOffsetContext previousOffset;
+    private final ConnectionContext connectionContext;
+    private final ReplicaSets replicaSets;
+    private final EventDispatcher<CollectionId> dispatcher;
+    protected final Clock clock;
+    private final SnapshotProgressListener snapshotProgressListener;
+    private final ErrorHandler errorHandler;
+    private AtomicBoolean aborted = new AtomicBoolean(false);
+
+    public MongoDbSnapshotChangeEventSource(MongoDbConnectorConfig connectorConfig, MongoDbTaskContext taskContext,
+                                            ReplicaSets replicaSets, MongoDbOffsetContext previousOffset,
+                                            EventDispatcher<CollectionId> dispatcher, Clock clock,
+                                            SnapshotProgressListener snapshotProgressListener, ErrorHandler errorHandler) {
+        this.connectorConfig = connectorConfig;
+        this.taskContext = taskContext;
+        this.connectionContext = taskContext.getConnectionContext();
+        this.previousOffset = previousOffset;
+        this.replicaSets = replicaSets;
+        this.dispatcher = dispatcher;
+        this.clock = clock;
+        this.snapshotProgressListener = snapshotProgressListener;
+        this.errorHandler = errorHandler;
+    }
+
+    @Override
+    public SnapshotResult execute(ChangeEventSourceContext context) throws InterruptedException {
```

Comment:
Again a common ancestor to this and `RelationalSnapshotChangeEventSource` would be good together with intorducing `doExectue()` method.

Reply:
See commit 0158b56.